### PR TITLE
【OSCP】在 SecretFlow 拆分学习中实现 max-norm 防御

### PR DESCRIPTION
**Pull request by Rainysponge**
**Type of change**
fixed: #1240
- [ ] Add new papers (Please tell us why you think this paper is awesome!)
- [ ] Fix the category of an existing paper/papers (Please tell us the reasons)
- [x] Add a new tool/primitive/application with a new markdown page (Thank you! Also, please tell us more about this awesome thing!)

**Description**

> This PR introduces the [max-norm defense](https://arxiv.org/pdf/2102.08504.) into SecretFlow. Maxnorm is a heuristic approach for protecting labels by matching the expected squared 2-norm of each perturbed gradient to the largest squared 2-norm in a mini-batch and adding specific noise only along the gradient direction.

**Result**
We tested the code using the Criteo dataset and the Wide & Deep model, with settings similar to those described in the paper. The result is shown as follows:

|              | Val AUC | Attack AUC |
|--------------|---------|------------|
| w/o max-norm | 77.88%  | 100.00%    |
| max-norm     | 54.75%  | 49.50%     |

### CODE REVIEW

File: secretflow/ml/nn/sl/defenses/max_norm.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    adding specific noise only along the gradient direction.
+    """
+
+    def __init__(self, backend: str = "torch", exec_device='cpu', **kwargs):
+        self.backend = backend.lower()
+        self.exec_device = exec_device
+        super().__init__(**kwargs)
+
+    def on_fuse_backward_end(self):
+        def average_gradient(worker):
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/max_norm.py:34]: 这个函数改个名吧，比如perturb_gradient

File: tests/ml/nn/sl/defenses/test_max_norm.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        train_label,
+        validation_data=(test_data, test_label),
+        epochs=1,
+        batch_size=128,
+        shuffle=False,
+        random_seed=1234,
+        callbacks=[max_norm],
+    )
+    print(history)
+    # attack_metrics = direction_lia.get_attack_metrics()
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/defenses/test_max_norm.py:180]: 没用的代码删一删

File: tests/ml/nn/sl/attack/model_def.py [hunk 3]
```diff
@@ -259,3 +261,260 @@ def forward(self, x):
...
+        inputs = X
+        for i in range(len(self.linear)):
+            fc = self.linear[i](inputs)
+            fc = self.activation(fc)
+            fc = self.dropout(fc)
+            inputs = fc
+        return inputs
+
+
+class WideDeepBottomAlice(nn.Module):
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/attack/model_def.py:301]: 同上一个pr

File: secretflow/ml/nn/sl/defenses/max_norm.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class perturb_gradient(Callback):
+    """
+    MaxNorm is a defense method designed to against the label inference attack
+    in Vertical Federated Learning and Split Learning: https://arxiv.org/pdf/2102.08504.
+    Maxnorm is a heuristic approach for protecting labels by matching the expected squared
+    2-norm of each perturbed gradient to the largest squared 2-norm in a mini-batch and
+    adding specific noise only along the gradient direction.
+    """
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/max_norm.py:26]: 参数注释同Cafe

File: secretflow/ml/nn/sl/defenses/max_norm.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        self.backend = backend.lower()
+        self.exec_device = exec_device
+        super().__init__(**kwargs)
+
+    def on_fuse_backward_end(self):
+        def average_gradient(worker):
+
+            gradient = worker._gradient
+            if self.backend == "tensorflow":
+                import tensorflow as tf
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/max_norm.py:38]: 同 Cafe

File: secretflow/ml/nn/sl/defenses/max_norm.py (hunk 1 above)
Comment (Rainysponge) [reply secretflow/ml/nn/sl/defenses/max_norm.py]: 已完成修改

**Review by zhaocaibei123 — DISMISSED** [review]

**Review by ian-huu — APPROVED** [review]
LGTM

**Review by zhaocaibei123 — APPROVED** [review]

**Review by ian-huu — APPROVED** [review]
